agentendpoint/agentendpoint.go

```diff
@@ -269,30 +289,73 @@ func (c *Client) loadTaskFromState(ctx context.Context) error {
 	return nil
 }
 
-// WaitForTaskNotification waits for and acts on any task notification indefinitely.
-func (c *Client) WaitForTaskNotification(ctx context.Context) error {
+func (c *Client) waitForTask(ctx context.Context) error {
+	stream, err := c.receiveTaskNotification(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = c.handleStream(ctx, stream)
+	if err == io.EOF {
+		// Server closed the stream indication we should reconnect.
+		return nil
+	}
+	if s, ok := status.FromError(err); ok && s.Code() == codes.Unavailable {
+		// Something canceled the stream (could be deadline/timeout), we should reconnect.
+		return nil
+	}
+	// TODO: Add more error checking (handle more API erros vs non API errors) and backoff where appropriate.
+	return err
+}
+
+// WaitForTaskNotification waits for and acts on any task notification until the Client is closed.
+// Multiple calls to WaitForTaskNotification will not create new watchers.
+func (c *Client) WaitForTaskNotification(ctx context.Context) {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+	if c.cancel != nil {
+		// WaitForTaskNotification is already running on this client.
+		return
+	}
+	logger.Debugf("Running WaitForTaskNotification")
+	ctx, c.cancel = context.WithCancel(ctx)
+
 	logger.Debugf("Checking local state file for saved task.")
 	if err := c.loadTaskFromState(ctx); err != nil {
 		logger.Errorf(err.Error())
 	}
 
 	logger.Debugf("Setting up ReceiveTaskNotification stream watcher.")
-	noti := make(chan struct{}, 1)
-	for {
-		token, err := config.IDToken()
-		if err != nil {
-			return fmt.Errorf("error fetching Instance IDToken: %v", err)
-		}
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				// We have been canceled.
+				logger.Debugf("Disabling WaitForTaskNotification")
+				return
+			default:
+			}
 
-		if err := c.receiveTaskNotification(ctx, noti, token); err != nil {
-			if s, ok := status.FromError(err); ok && s.Code() == codes.PermissionDenied {
-				// Service is not enabled for this project.
-				time.Sleep(config.SvcPollInterval())
-				continue
+			if err := c.waitForTask(ctx); err != nil {
```

Comment:
Currently the error handling is divided between two methods for stream failures, is that on purpose?
You should ideally handle all stream errors in WaitForTask and based on the logic of retrying/swallowing up error should pass a known exception to the WaitForTaskNotification method. Does that sound reasonable?

Reply:
Yes, handling it all in waitForTask makes the testing much simpler too. Really all i need is an error for the long sleep right now (PermissionDenied)